File: packages/ios/Sources/Components/CategoryField/CategoryField.swift
import SwiftUI

/// The search field's form once a quick-access category is chosen — the
/// prototype's, measured: a 48-tall, control-radius field in the category's
/// colour at 12 % with a 1-point border of it, the category's icon at 28 in
/// the colour, its name at 15 semibold in the foreground, a 22-tall count
/// pill filled with the colour, and a 32 clear at the trailing edge with its
/// cross in the foreground. It takes the field's place in the search row, so
/// its width is the row's to give. The name and the cross are in the
/// foreground because the category colour on its own wash fails 4.5:1 for
/// seven of the eight tints (Olcay, 2026-09-21); the icon is decorative, the
/// name says what it shows.
///
/// Mirrors the React `CategoryField`.
public struct KozmosCategoryField<Icon: View>: View {
    private let label: String
    private let count: Int?
    /// The count's spoken form; the host pluralises and localises it.
    private let countLabel: (Int) -> String
    private let tint: KozmosCategoryTint
    private let clearLabel: String
    private let onClear: () -> Void
    private let icon: Icon

    public init(
        label: String,
        count: Int? = nil,
        countLabel: @escaping (Int) -> String = { "\($0) places" },
        tint: KozmosCategoryTint = KozmosCategoryTint(accent: KozmosColors.primitivesColorsTheme500, fill: KozmosInkedFill(fill: KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle, ink: KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle)),
        clearLabel: String = "Clear category",
        onClear: @escaping () -> Void,
        @ViewBuilder icon: () -> Icon
    ) {
        self.label = label
        self.count = count
        self.countLabel = countLabel
        self.tint = tint
        self.clearLabel = clearLabel
        self.onClear = onClear
        self.icon = icon()
    }

    static var height: CGFloat { 48 }
    static var iconSize: CGFloat { 28 }
    static var pillHeight: CGFloat { 22 }
    static var clearSize: CGFloat { 32 }

    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
            icon
                .frame(width: Self.iconSize, height: Self.iconSize)
                .foregroundColor(tint.accent)
                .accessibilityHidden(true)
            Text(label)
                .font(.system(size: 15, weight: .semibold))
                .foregroundColor(KozmosColors.primitivesColorsForeground0)
                .lineLimit(1)
            if let count {
                Text("\(count)")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundColor(tint.fill.ink)
                    .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing75)
                    .frame(minWidth: Self.pillHeight, minHeight: Self.pillHeight)
                    .background(Capsule().fill(tint.fill.fill))
                    .accessibilityLabel(countLabel(count))
            }
            Spacer(minLength: 0)
            Button(action: onClear) {
                Image(systemName: "xmark")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(KozmosColors.primitivesColorsForeground0)
                    .frame(width: Self.clearSize, height: Self.clearSize)
                    .contentShape(Circle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(clearLabel)
        }
        .padding(.leading, KozmosDimensions.primitivesLayoutSpacing150)
        .padding(.trailing, KozmosDimensions.primitivesLayoutSpacing100)
        .frame(height: Self.height)
        .background(tint.accent.opacity(0.12))
        .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                .strokeBorder(tint.accent, lineWidth: 1)
        )
        .accessibilityElement(children: .contain)
        .accessibilityLabel(count.map { "\(label), \(countLabel($0))" } ?? label)
    }
}

File: packages/android/src/main/java/com/kozmos/components/CategoryField/CategoryField.kt
package com.kozmos.components.categoryfield

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.LocalIndication
import androidx.compose.foundation.clickable
import androidx.compose.foundation.indication
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kozmos.components.categorytile.KozmosCategoryTint
import com.kozmos.components.counter.KozmosInkedFill
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/**
 * The search field's form once a quick-access category is chosen — the
 * prototype's, measured: 48 tall, the control radius, the category's colour
 * at 12 % with a 1-dp border of it, the icon at 28, the name at 15 semibold,
 * a 22-tall count pill filled with the colour, a 32 clear at the end. It
 * takes the field's place in the search row.
 *
 * Mirrors `CategoryField` on the web and `KozmosCategoryField` on iOS.
 */
@Composable
fun KozmosCategoryField(
    label: String,
    onClear: () -> Unit,
    modifier: Modifier = Modifier,
    count: Int? = null,
    tint: KozmosCategoryTint = KozmosCategoryTint(
        KozmosColors.primitivesColorsTheme500,
        KozmosInkedFill(KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle, KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle)
    ),
    clearLabel: String = "Clear category",
    countLabel: (Int) -> String = { "$it places" },
    icon: (@Composable () -> Unit)? = null
) {
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
    Row(
        modifier = modifier
            .height(48.dp)
            .clip(shape)
            .background(tint.accent.copy(alpha = 0.12f))
            .border(1.dp, tint.accent, shape)
            .padding(start = KozmosDimensions.primitivesLayoutSpacing150, end = KozmosDimensions.primitivesLayoutSpacing25)
            .semantics(mergeDescendants = false) {
                contentDescription = if (count == null) label else "$label, ${countLabel(count)}"
            },
        verticalAlignment = Alignment.CenterVertically
    ) {
        CompositionLocalProvider(LocalContentColor provides tint.accent) {
            if (icon != null) {
                // Decorative: the name beside it says what it shows.
                Box(
                    modifier = Modifier.size(28.dp).clearAndSetSemantics {},
                    contentAlignment = Alignment.Center
                ) { icon() }
                Spacer(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSpacing100))
            }
            // The name and its count take what the icon and the clear leave:
            // the name shrinks first, the pill follows it, and the clear stays
            // at the trailing edge, as on React and iOS. A spacer weighted
            // against the weighted, non-filling name left the name's unused
            // share after the clear, so the clear floated inward (2026-09-21).
            Row(modifier = Modifier.weight(1f), verticalAlignment = Alignment.CenterVertically) {
                // The name and the clear's cross in the foreground, themed:
                // the category colour on its own wash failed 4.5:1 for seven
                // tints (Olcay, 2026-09-21).
                Text(
                    text = label,
                    color = KozmosThemeTokens.primitivesColorsForeground0,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f, fill = false)
                )
                if (count != null) {
                    Spacer(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSpacing100))
                    Box(
                        modifier = Modifier
                            .defaultMinSize(minWidth = 22.dp, minHeight = 22.dp)
                            .clip(CircleShape)
                            .background(tint.fill.fill)
                            .padding(horizontal = KozmosDimensions.primitivesLayoutSpacing75)
                            .semantics { contentDescription = countLabel(count) },
                        contentAlignment = Alignment.Center
                    ) {
                        Text(text = count.toString(), color = tint.fill.ink, fontSize = 12.sp, fontWeight = FontWeight.SemiBold)
                    }
                }
            }
            // The clear: a 32 circle to see, the 44 target around it to hit,
            // as the search bar's (Olcay, 2026-09-21). The row's end padding
            // is 2, so the circle sits 8 from the edge; the press shows on
            // the circle.
            val clearInteraction = remember { MutableInteractionSource() }
            Box(
                modifier = Modifier
                    .size(44.dp)
                    .clickable(
                        interactionSource = clearInteraction,
                        indication = null,
                        role = Role.Button,
                        onClick = onClear
                    )
                    .semantics { contentDescription = clearLabel },
                contentAlignment = Alignment.Center
            ) {
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .indication(clearInteraction, LocalIndication.current),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(imageVector = Icons.Default.Close, contentDescription = null, tint = KozmosThemeTokens.primitivesColorsForeground0, modifier = Modifier.size(16.dp))
                }
            }
        }
    }
}
